src/rdf4cpp/rdf/Serialize.hpp

```diff
@@ -0,0 +1,220 @@
+
+    { ser.buffer() } -> std::same_as<typename T::Buffer &>;
+    { ser.cursor() } -> std::same_as<Cursor &>;
+    { ser.finalize() } -> std::convertible_to<typename T::Output>;
+    T::flush(&ser.buffer(), ser.cursor());
+};
+
+/**
+ * Flushes user defined data
+ */
+using FlushFunc = void (*)(void *buffer, Cursor &cursor) noexcept;
+
+/**
+ * Serializes the given string as is
+ * See Node::serialize for more details on the signature/usage
+ */
+inline bool serialize_str(std::string_view str, void *const buffer, Cursor &cursor, FlushFunc const flush) noexcept {
+    while (true) {
+        auto const max_write = std::min(str.size(), cursor.size());
+        memcpy(cursor.data(), str.data(), max_write);
+        cursor.advance(max_write);
+
+        if (max_write == str.size()) [[likely]] {
+            break;
+        } else {
+            if (flush(buffer, cursor); cursor.size() == 0) [[unlikely]] {
+                return false;
+            }
+
+            str.remove_prefix(max_write);
+        }
+    }
+    return true;
+}
+
+/**
+ * Serializes a string as is
+ * See Node::serialize<> for more details on the signature/usage
+ */
+template<Serializer Ser>
+bool serialize_str(std::string_view str, Ser &ser) {
+    return serialize_str(str, &ser.buffer(), ser.cursor(), &Ser::flush);
+}
+
+/**
+ * (Optional) base class for serializers.
+ * Encapsulates a cursor and some buffer type and makes the implementation of flush less error prone.
+ */
+template<typename CRTP, typename Buffer>
+struct SerializerBase {
+private:
+    Buffer buffer_;
+    Cursor cursor_;
+
+public:
+    template<typename ...BufferArgs>
+    explicit constexpr SerializerBase(BufferArgs &&...buffer_args) : buffer_{std::forward<BufferArgs>(buffer_args)...} {
+    }
+
+    [[nodiscard]] constexpr Cursor &cursor() noexcept { return cursor_; }
+    [[nodiscard]] constexpr Buffer &buffer() noexcept { return buffer_; }
+
+    static void flush(void *buffer, Cursor &cursor) noexcept {
+        CRTP::flush_impl(*reinterpret_cast<typename CRTP::Buffer *>(buffer), cursor);
+    }
+};
+
+using StringBuffer = std::string;
+
+/**
+ * A serializer that serializes to a std::string
+ */
+struct StringSerializer : SerializerBase<StringSerializer, StringBuffer> {
+public:
+    using Buffer = StringBuffer;
+    using Output = std::string;
+
+    explicit StringSerializer(size_t cap = 256) noexcept {
+        buffer().resize(cap);
+        cursor().repoint(buffer().data(), buffer().size());
+    }
+
+    [[nodiscard]] Output &finalize() noexcept {
+        buffer().resize(static_cast<size_t>(cursor().data() - buffer().data()));
+        return buffer();
+    }
+
+    static void flush_impl(Buffer &buffer, Cursor &cursor) noexcept {
+        auto const bytes_written = cursor.data() - buffer.data();
+
+        buffer.resize(buffer.size() * 2);
+        cursor.repoint(buffer.data() + bytes_written,
+                       cursor.size() + buffer.size());
+    }
+};
+
+struct CFileBuffer {
```

Comment:
I mean, its not really since you have to pass in `FILE *`. I would also keep it that way because then error handling can be done however the user wants.